src/stan/model/indexing/lvalue.hpp

```diff
@@ -253,16 +251,17 @@ assign(Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic>& x,
  * @param[in] depth Indexing depth (default 0).
  * @throw std::out_of_range If either of the indices are out of bounds.
  */
-template <typename T, typename U>
-void assign(Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic>& x,
+template <typename EigMat, typename U, typename = require_eigen_t<EigMat>,
+          typename = require_not_eigen_vector_t<EigMat>>
+void assign(EigMat& x,
             const cons_index_list<
-                index_uni, cons_index_list<index_uni, nil_index_list> >& idxs,
+                index_uni, cons_index_list<index_uni, nil_index_list>>& idxs,
             const U& y, const char* name = "ANON", int depth = 0) {
```

Comment:
I think if a function takes in generic scalars it should be by reference. If a `var` is brough in by value that will cause another vari to go on the stack right?

Reply:
I agree that generic scalars should be passed by reference.  Constant reference, in fact, because that allows constants.

But it's not because new `vari` go on the stack.  You only get a new stack entry with a new value.

```
var x;  // x.vi_ == nullptr
x = 12;  // x.vi_ points to new vari on stack with value 12
var y = x;  // x.vi_ == y.vi_ (no new pointer allocation)
```

Weren't you suggesting passing `var` by value?  It's still only a pointer in size (no vtable), so that'd make sense.  Once you get to fvar<T> for scalar T it's two pointers, and fvar<fvar<T>> is four.